stan/math/prim/meta/apply_scalar_binary.hpp

```diff
@@ -0,0 +1,326 @@
+#include <vector>
+
+namespace stan {
+namespace math {
+
+// Forward declaration to allow specialisations
+template <typename T1, typename T2, typename Enable = void,
+          typename Enable2 = void>
+struct apply_scalar_binary {};
+
+/**
+ * Base template class for vectorization of binary scalar functions
+ * defined by applying a functor to a combination of scalars,
+ * containers of matching sizes, or a combination of a scalar and a container.
+ * These containers can be a standard library vector, Eigen dense
+ * matrix expression template, or container of these. For each specialisation,
+ * the same type as the largest (dimension) input is returned.
+ *
+ * This base template class takes (and returns) scalars.
+ */
+template <typename T1, typename T2>
+struct apply_scalar_binary<T1, T2, require_all_stan_scalar_t<T1, T2>> {
+  /**
+   * Member function for applying a functor to two scalars and subsequently
+   * returning a scalar.
+   *
+   * @tparam T1 Type of first argument to which functor is applied.
+   * @tparam T2 Type of second argument to which functor is applied.
+   * @tparam F Type of functor to apply.
+   * @param x First input to which operation is applied.
+   * @param y Second input to which operation is applied.
+   * @param f functor to apply to inputs.
+   * @return Scalar with result of applying functor to input.
+   */
+  template <typename F>
+  static inline auto apply(const T1& x, const T2& y, const F& f) {
+    return f(x, y);
+  }
+};
+
+/**
+ * Specialisation for use with two Eigen inputs. Eigen's binaryExpr framework
+ * is used for more efficient indexing of both row- and column-major inputs
+ * without separate loops.
+ */
+template <typename T1, typename T2>
+struct apply_scalar_binary<T1, T2, require_all_eigen_t<T1, T2>> {
+  /**
+   * Member function for applying a functor to two Eigen objects and
+   * subsequently returning an Eigen object.
+   *
+   * @tparam T1 Type of first argument to which functor is applied.
+   * @tparam T2 Type of second argument to which functor is applied.
+   * @tparam F Type of functor to apply.
+   * @param x First Eigen input to which operation is applied.
+   * @param y Second Eigen input to which operation is applied.
+   * @param f functor to apply to Eigen input.
+   * @return Eigen object with result of applying functor to inputs.
+   */
+  template <typename F>
+  static inline auto apply(const T1& x, const T2& y, const F& f) {
+    return x.binaryExpr(y, f).eval();
+  }
+};
+
+/**
+ * Specialisation for use with one Eigen input and one scalar. Eigen's
+ * unaryExpr framework is used for more efficient indexing of both row-
+ * and column-major inputs. The unaryExpr framework also allows for the
+ * scalar to be captured and applied to each element in the Eigen object.
+ */
+template <typename T1, typename T2>
+struct apply_scalar_binary<T1, T2, require_any_eigen_t<T1, T2>,
+                           require_any_stan_scalar_t<T1, T2>> {
```

Comment:
How did you decide when to speciallize the `apply_scalar_binary` struct and when to overload the `apply` function in the struct? I would like one of those options used consistently. 

Actually, since you use requires to select the overload, we don't need a struct at all. You could just overload a free function. That would also let you save a few lines on docs.

Reply:
Good idea, will do